mindmeld/components/dialogue.py

```diff
@@ -326,15 +332,96 @@ def apply_handler(self, request, responder, target_dialogue_state=None):
         if self.async_mode:
             return self._apply_handler_async(
                 request, responder, target_dialogue_state=target_dialogue_state)
+
+        return self._apply_handler_sync(
+            request, responder, target_dialogue_state=target_dialogue_state)
+
+    def _apply_handler_sync(self, request, responder, target_dialogue_state=None):
+        """Applies the dialogue state handler for the most complex matching rule
+
+         Args:
+            request (Request): The request object.
+            responder (DialogueResponder): The responder object.
+            target_dialogue_state (str, optional): The target dialogue state.
+
+        Returns:
+            (dict): A dict containing the dialogue state and directives.
+        """
+        for _ in range(3):
+            try:
+                return self._attempt_handler_sync(
+                    request, responder, target_dialogue_state=target_dialogue_state
+                )
+            except DialogueStateException as e:
+                if e.target_dialogue_state and e.target_dialogue_state != target_dialogue_state:
+                    target_dialogue_state = e.target_dialogue_state
+                else:
+                    self.logger.warning(
+                        "Ignoring target dialogue state '{}'".format(e.target_dialogue_state)
+                    )
+                    target_dialogue_state = None
+
+        if target_dialogue_state:
+            self.logger.warning(
+                "Ignoring target dialogue state '{}'".format(target_dialogue_state)
+            )
+        return self._attempt_handler_sync(request, responder)
+
+    def _attempt_handler_sync(self, request, responder, target_dialogue_state=None):
+        """Tries to apply the dialogue state handler for the most complex matching rule
+
+         Args:
+            request (Request): The request object.
+            responder (DialogueResponder): The responder object.
+            target_dialogue_state (str, optional): The target dialogue state.
+
+        Returns:
+            (dict): A dict containing the dialogue state and directives.
+        """
         dialogue_state = self._get_dialogue_state(request, target_dialogue_state)
         handler = self._get_dialogue_handler(dialogue_state)
         responder.dialogue_state = dialogue_state
-        handler(request, responder)
-        return responder
+        res = handler(request, responder)
+
+        if res and 'dialogue_state' in res:
+            # Add dialogue flow's sub-dialogue_state if provided
+            dialogue_state = '.'.join([dialogue_state, res["dialogue_state"]])
+        return {'dialogue_state': dialogue_state, 'directives': responder.directives}
```

Comment:
We are returning a responder object for the apply_handler function, but here we are returning a dict. Please fix this.

Reply:
Done.